### ISSUE TITLE
[inductor] [cuda] [fake tensor] `torch.ones(x.size(0))` becomes a fake tensor for `torch.diagonal_scatter`

### ISSUE THREAD

**Issue by shaoyuyoung** [pytorch/pytorch#151670]
### 🐛 Describe the bug

**symptom**: `torch.diagonal_scatter` throws fake tensor error, misaligning with eager.
**device**: only cuda
**repro**
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch._inductor import config

config.fallback_random = True
torch.set_grad_enabled(False)
import os

os.environ['TORCHDYNAMO_VERBOSE'] = '1'


class Model(nn.Module):

    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        y = torch.ones(x.size(0))
        x = torch.diagonal_scatter(x, y)
        return x


model = Model()

x = torch.rand(1, 2)

inputs = [x]


def run_test(model, inputs, device, backend):
    torch.manual_seed(0)
    model = model.to(device)
    inputs = [x.to(device) for x in inputs]
    if backend != "eager":
        model = torch.compile(model, backend=backend)
    try:
        output = model(*inputs)
        print(f"succeed on {backend}")
    except Exception as e:
        print(e)


run_test(model, inputs, 'cuda', 'eager')
run_test(model, inputs, 'cuda', 'inductor')
```

### Error logs

```
succeed on eager
Dynamo failed to run FX node with fake tensors: call_function <built-in method diagonal_scatter of type object at 0x7f4ca6e5a2c0>(*(FakeTensor(..., device='cuda:0', size=(1, 2)), FakeTensor(..., size=(1,))), **{}): got RuntimeError('Unhandled FakeTensor Device Propagation for aten.diagonal_scatter.default, found two different devices cuda:0, cpu')
```

### Versions

nightly 20250414

**Comment by shaoyuyoung** [comment]
related to #144748
cc @zero000064 , mind taking a look at this?
thanks!

**Comment by zero000064** [comment]
fake tensor runs device check on all operations, https://github.com/pytorch/pytorch/blob/0b0da81021e061c021e515bc35d7dc0dbbb05941/torch/_subclasses/fake_tensor.py#L833, so only zero dim cpu tensors could get exemptions. Obviously, your case will be blocked by this check.
I didn't find diagonal_scatter in RegisterCUDAEverything.cpp, but my guess is device check not enforced in that opeartion.
If you check generated RegisterCUDAEverything.cpp in the build folder, only a subset of operations are enforced device check. Arithmetic operations don't have that check.

**Comment by shaoyuyoung** [comment]
Thanks for your comment!
Did you mean this is not case-specific? Many operators have the same fake tensor issue, except the arithmetic op (such as `torch.add`) (?)

**Comment by zero000064** [comment]
Yes, there're inconsistencies due to device check logic discrepancy.
https://github.com/pytorch/pytorch/pull/147501 is to use an allow list to align them, then this kind of issue could be fixed by expanding the allow list, but I'm not sure if it would go through or not.

**Comment by xgz2** [comment]
Hi @zou3519 , thanks for the great hint. Is test2 supposed to succeed in FakeTensorMode if it currently doesn't succeed in eager mode?
